templates/transit-gateway-attachment.yaml

```diff
@@ -0,0 +1,61 @@
+AWSTemplateFormatVersion: 2010-09-09
+Description: >-
+  Setup an AWS transit gateway attachment
+Parameters:
+  Department:
+    Description: 'The department for this resource'
+    Type: String
+    AllowedPattern: '^\S*$'
+    ConstraintDescription: 'Must be string with no spaces'
+    Default: 'Platform'
+  Project:
+    Description: 'The name of the project that this resource is used for'
+    Type: String
+    AllowedPattern: '^\S*$'
+    ConstraintDescription: 'Must be string with no spaces'
+    Default: 'Infrastructure'
+  OwnerEmail:
+    Description: 'Email address of the owner of this resource'
+    Type: String
+    AllowedPattern: '^(([^<>()\[\]\\.,;:\s@"]+(\.[^<>()\[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$'
+    ConstraintDescription: 'Must be an acceptable email address syntax(i.e. joe.smith@sagebase.org)'
+    Default: 'it@sagebase.org'
+  TransitGatewayId:
+    Description: 'The transit gateway Id'
+    Type: String
+    AllowedPattern: '^tgw-[a-z0-9]+$'
+    ConstraintDescription: 'Must be a transit gateway Id (i.e. tgw-1111148306e2d3b5)'
+    Default: 'tgw-03447748306e2d3b5'   # ID for sage-tgw transit gateway
+  VpcName:
+    Description: 'The transit gateway will attach to this VPC Id'
+    Type: String
+Resources:
+  TransitGatewayAttachment:
+    Type: AWS::EC2::TransitGatewayAttachment
+    Properties:
+      SubnetIds:
+        - !ImportValue
+          'Fn::Sub': '${AWS::Region}-${VpcName}-PrivateSubnet'
+        - !ImportValue
+          'Fn::Sub': '${AWS::Region}-${VpcName}-PrivateSubnet1'
+        - !ImportValue
+          'Fn::Sub': '${AWS::Region}-${VpcName}-PrivateSubnet2'
```

Comment:
Are there any use cases for public subnets? maybe routing to a service through the transit gateway but not through the internet?

Reply:
Yes.  however AWS attachments only allows a single AZ to single subnet mapping.  We designed our VPCs with multiple subnets in a single zone.   For example, the "PublicSubnet1" and "PrivateSubnet1" both are in "us-east-1a".  This means that we can attach either "PublicSubnet1" OR "PrivateSubnet1" to "us-east-1a" but not both.  I decided that there are more situations where we want to route to private subnets than public ones which is why I've set it up like this. I've never run into this restriction with any other AWS service.  It might be worth looking into refactoring our VPC architecture so that every subnet is in a separate zone.